### ISSUE TITLE
Randomly cannot login with OTP 2fa anymore (urgent)

### ISSUE THREAD

**Issue by tech4242** [gravitational/teleport#13338]
Expected behavior: login seamlessly with OTP with existing users

Current behavior: 
Randomly login does not work - I am currently login in the UI in the browser but I cannot `tsh login` I try over and over again and it always says `ERROR: invalid username, password or second factor`. I tried with another user and he has the same issue. He can login to the browser but not with the terminal. 

Also tried deleting his user and then randomly `invalid one time token, please check if the token has expired and try again` happens but eventually just by trial and error the Web UI got it.  Same thing on reset password - had to enter multiple OTP codes to get it right.

We can still both use our OTP apps for various other services and we have not performed any Teleport updates.

Bug details:
- Teleport version: v9.1.0
- Recreation steps: unclear
- Debug logs: what would you like to have?

**Comment by zmb3** [comment]
Please share auth service logs around the time when the login is denied.

**Comment by tech4242** [comment]
hey @zmb3 not exactly sure what to give you but here is an excerpt from the audit sessions from the UI from my own last failed login attempt from my personal terminal:

```json
{
  "cluster_name": "tp.xx.xxx",
  "code": "T1000W",
  "ei": 0,
  "error": "invalid username, password or second factor",
  "event": "user.login",
  "method": "local",
  "success": false,
  "time": "2022-06-09T15:33:27.029Z",
  "uid": "xx",
  "user": "teleport-xxx"
}
```

**Comment by tech4242** [comment]
I can still open a root session with the machine itself and use `tctl` through the UI but absolutely no chance logging in with tsh login.

We have all the default config in our teleport.yaml and cap.yaml seems ok as well. can also add them here

**Comment by tech4242** [comment]
maybe some more info from `systemctl status teleport.service`:

```bash
teleport.service - Teleport SSH Service
     Loaded: loaded (/lib/systemd/system/teleport.service; disabled; vendor preset: enabled)
     Active: active (running) since Fri 2022-04-22 13:22:48 UTC; 1 months 17 days ago
   Main PID: 3888 (teleport)
      Tasks: 21 (limit: 4612)
     Memory: 136.6M
     CGroup: /system.slice/teleport.service
             ├─  3888 /usr/local/bin/teleport start --pid-file=/run/teleport.pid
             ├─406913 /usr/local/bin/teleport exec
             ├─406920 -bash
             ├─407629 systemctl status teleport.service
             └─407630 pager

Jun 09 15:46:43 xxx teleport[3888]:         /go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:344 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).handleConn
Jun 09 15:46:43 xxx teleport[3888]:         /go/src/github.com/gravitational/teleport/lib/srv/alpnproxy/proxy.go:282 github.com/gravitational/teleport/lib/srv/alpnproxy.(*Proxy).Serve.func1
Jun 09 15:46:43 xxx teleport[3888]:         /opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
Jun 09 15:46:43 xxxx teleport[3888]: User Message: acme/autocert: missing server name] alpnproxy/proxy.go:290
Jun 09 15:46:43 xxxx teleport[3888]: 2022-06-09T15:46:43Z WARN [WEB]       http: named cookie not present request:GET /web/reset/88272ed913d4b67c7d4fa64e7a39faee web/apiserver.go:2592
Jun 09 15:50:09 xxxx teleport[3888]: 2022-06-09T15:50:09Z INFO [CA]        Generating TLS certificate {0x84debd0 0xc001760180 1.3.9999.1.7=#130d74702e6b72616e75732e646576,1.3.9999.1.3=#13136b68632d6f74632d6465762d636c757374657>
Jun 09 15:50:09 xxx teleport[3888]: 2022-06-09T15:50:09Z INFO [AUDIT]     cert.create cert_type:user cluster_name:xxx code:TC000I ei:0 event:cert.create identity:map[expires:2022-06-10T02:46:35.074788519Z kubernetes>
Jun 09 15:51:46 xxx teleport[3888]: 2022-06-09T15:51:46Z INFO [CA]        Generating TLS certificate {0x84debd0 0xc001f11750 1.3.9999.1.7=#130d74702e6b72616e75732e646576,1.3.9999.1.3=#13136b68632d6f74632d6465762d636c757374657>
Jun 09 15:51:46 xxxteleport[3888]: 2022-06-09T15:51:46Z INFO [AUDIT]     cert.create cert_type:user cluster_name:xxx code:TC000I ei:0 event:cert.create identity:map[expires:2022-06-10T03:15:31.025791411Z kubernetes>
Jun 09 15:52:49 xxxx teleport[3888]: 2022-06-09T15:52:49Z INFO [AUDIT]     user.login cluster_name:xxx code:T1000W ei:0 error:[too many incorrect attempts, please try again later] event:user.login method:local succes>
```

**Comment by zmb3** [comment]
To be able to help, we would need to see more of the logs from your Teleport Auth Server when the connection is denied. 

If you're using systemd, `journalctl -u teleport` should dump them.

**Comment by zmb3** [comment]
Did your system clock change in some way or otherwise drift?

**Comment by tech4242** [comment]
it's theoretically possible but I haven't touched the VM (Ubuntu 20.4) since the initial setup. 

do you mean like a GMT thing with the timestamps or something else? anyway to corroborate this?

**Comment by tech4242** [comment]
it def. worked fine for 2-3 weeks

**Comment by zmb3** [comment]
> do you mean like a GMT thing with the timestamps or something else? anyway to corroborate this?

In order for TOTP to work, the time between the client where you're logging in and the server where you're validating the OTP must be in sync. Can you check the time on both systems for drift?

See also #1727

**Comment by tech4242** [comment]
hmm I see also explains the error message: https://security.stackexchange.com/questions/173161/does-the-totp-algorithm-rely-on-the-client-time-always-being-synced-correctly

I waited for the token to expire, another 10 seconds & then tried and it worked. I guess you are right. I see there are some strategies to actually prevent this clock drift issue but none are implemented here. Explains why it worked in the browser after several tries - the UI doesn't block access requests as quickly as the terminal CLI so the user didn't get blocked after multiple login attempts with same token (I just pressed button a few times out of frustration : D)

how would we go about having a good fix here beyond waiting for a token to expire and hope the drift is gone?
